exaudfclient/base/javacontainer/script_options/parser_ctpg.cc

```diff
@@ -1,9 +1,147 @@
+    } catch(const ExecutionGraph::OptionParserException& ex) {
+        Utils::rethrow(ex, "F-UDF-CL-SL-JAVA-1625");
+    }
+}
+
+void ScriptOptionLinesParserCTPG::extractImportScripts(SwigFactory & swigFactory) {
+
+    try {
+        parse();
+    } catch(const ExecutionGraph::OptionParserException& ex) {
+        Utils::rethrow(ex, "F-UDF-CL-SL-JAVA-1626");
+    }
+
+    const auto optionIt = m_foundOptions.find(m_keywords.importKeyword());
+    if (optionIt != m_foundOptions.end()) {
+        CTPG::ScriptImporter scriptImporter(swigFactory, m_keywords);
+        scriptImporter.importScript(m_scriptCode, m_foundOptions);
+        //The imported scripts will change the location of the other options in m_foundOptions
+        //Also there might be new JVM / External Jar options
+        //=> We need to clear the option map and reset the parser.
+        m_foundOptions.clear();
+        m_needParsing = true;
+    }
+}
+
+std::string && ScriptOptionLinesParserCTPG::getScriptCode() {
+    try {
+        parse();
+    } catch(const ExecutionGraph::OptionParserException& ex) {
+        Utils::rethrow(ex, "F-UDF-CL-SL-JAVA-1627");
+    }
+    //Remove all options from script code in reverse order
+    struct option_location {
+        size_t pos;
+        size_t len;
+    };
+    struct comp {
+        bool operator()(option_location a, option_location b) const {
+            return a.pos > b.pos;
+        }
+    };
+    std::set<option_location, comp> option_locations;
+    for (const auto & option: m_foundOptions) {
+        for (const auto & option_loc: option.second) {
+            option_location loc = { .pos = option_loc.idx_in_source, .len = option_loc.size};
+            option_locations.insert(loc);
+        }
+    }
+    for (const auto option_loc: option_locations) {
+        m_scriptCode.erase(option_loc.pos, option_loc.len);
+    }
+    return std::move(m_scriptCode);
+}
+
+void ScriptOptionLinesParserCTPG::parse() {
+    if (m_needParsing) {
+        if(!m_foundOptions.empty()) {
+            throw std::logic_error("F-UDF-CL-SL-JAVA-1620 Internal error. Parser result is not empty.");
+        }
+        try {
+            ExecutionGraph::OptionsLineParser::CTPG::parseOptions(m_scriptCode, m_foundOptions);
+        } catch(const ExecutionGraph::OptionParserException& ex) {
+            Utils::rethrow(ex, "F-UDF-CL-SL-JAVA-1621");
+        }
+
+        m_needParsing = false;
+
+        //Check for unknown options
+        for (const auto & option: m_foundOptions) {
+            if (m_keywords.jarKeyword() != option.first &&
+                m_keywords.scriptClassKeyword() != option.first  &&
+                m_keywords.importKeyword() != option.first  &&
+                m_keywords.jvmKeyword() != option.first) {
+                    std::stringstream ss;
+                    ss << "F-UDF-CL-SL-JAVA-1622 " << "Unexpected option: " << option.first;
+                    throw std::invalid_argument(ss.str());
```

Comment:
Hm, is this really a exception, with old one the option simply would be ignored and stays in the text

Reply:
I could have implemented the same behavior as in the old legacy parser, but then I would need to filter the known options in ScriptOptionLinesParserCTPG::getScriptCode() (where we remove the options from script code).

My thought was to keep the implementation simple, and also throwing an explicit error here is better than getting a strange syntax error back from the Java compiler later.

Reply:
Oh, you are right we know the language and that every option line that we encounter will be wrong.